### DISCUSSION TITLE
php 8.1 and above

### DISCUSSION THREAD

**Discussion by geotronik** [koseven/koseven#524] [Q&A]
I just found this koseven version of kohana and am pretty happy!  
I would like to know if there is anything above 8.0 in the works. 
I have php 8.030 running fine with the information I found here but anything above 8.0 gives some errors

Sorry in advance if I am not following some protocol to ask for things here. I will be donating to this project if I can get this project I've had running for a few years. But i want to have this go to a newer environment: namely alma 9, postgres 16, php 8.1+

`Fatal error: Uncaught Error: Class "HTTP_Header" not found in C:\webdocs\system\classes\Kohana\Response.php:158 
Stack trace: #0 C:\webdocs\system\classes\Kohana\Response.php(31): Kohana_Response->__construct(Array)
 #1 C:\webdocs\system\classes\Kohana\Kohana\Exception.php(272): Kohana_Response::factory() 
#2 C:\webdocs\system\classes\Kohana\Kohana\Exception.php(110): Kohana_Kohana_Exception::response(Object(ErrorException)) 
#3 C:\webdocs\system\classes\Kohana\Kohana\Exception.php(86): Kohana_Kohana_Exception::_handler(Object(ErrorException)) 
#4 C:\webdocs\system\classes\Kohana\Core.php(986): Kohana_Kohana_Exception::handler(Object(ErrorException)) 
#5 [internal function]: Kohana_Core::shutdown_handler() 
#6 {main} thrown in C:\webdocs\system\classes\Kohana\Response.php on line 158`

**Comment by uBizzy** [comment]
@geotronik, I am wotking on this. This is related to functions type definitions. Hope to have most of that solved soon.

**Comment by geotronik** [comment]
FYI, 
I found a solution for most of my issues with php8.2+. in Kohana with Koseven updates.  I believe it would be fine with 8.3 as well.

Anything having to do with ArrayObject needs to be set with the following attribute when functions return something different than expected, I think is the best way to explain it.

#[\ReturnTypeWillChange]
https://www.php.net/manual/en/class.returntypewillchange.php

For the files:
system/classes/Kohana/Config/Group.php
system/classes/Cookie.php
system/classes/Core.php 
system/classes/HTTP/Header.php

I added the attribute 
#[\ReturnTypeWillChange]
above each function call that has an issue revealing the same basic error I mentioned above.

This specific example below occurs in the system/classes/Kohana/Config/Group.php
```php
        #[\ReturnTypeWillChange]
	public function offsetSet($key, $value)
	{
		$this->_parent_instance->_write_config($this->_group_name, $key, $value);
		return parent::offsetSet($key, $value);
	}
``` 
Once I applied this attribute to each function as I debugged each operation that used it I have a 100% working app that migrated from php7 to 8.2.12. regarding Kohana base functionality.

**Reply by uBizzy** [reply]
@geotronik , this is not a good option. It will just mute the output error. These errors are related to their parent definitions. So you just need to find what the return definition is set on their parent.

For eample. The "Group" class on kohana/koseven extends "ArrayObject" php's internal class. Checking this particular method "offsetSet()" you have the following definition:

```
public offsetSet(mixed $key, mixed $value): void
```

So the kohana/koseven method should declare the correct return type. On this case is "void". Then we should fix this method to the following on kohana/koseven:

```
public function offsetSet($key, $value) : void
```

**Reply by geotronik** [reply]
I did kind of  realize this is not a proper solution. But least I can temporarily get this critical site moving forward until an update is made that will replace this way of solving my issues.  I will be logging this in an upgrade.log so i know where too look when and if it is resolved properly.

**Comment by phpclub** [comment]
Try (unstable)
https://github.com/koseven/koseven/tree/php8.3

**Comment by timhj** [comment]
For the morepart, https://github.com/koseven/koseven/tree/devel works with PHP 8.1. 

For PHP 8.4, merging the changes from [@worateps php 8.4](https://github.com/worateps/koseven/tree/feature/php8.4-compatibility) into the devel branch (there were a couple of conflicts to resolve but nothing major) was a success and the merged code runs happily on PHP 8.4.

The [PHP 8.3](https://github.com/koseven/koseven/tree/php8.3) branch had a lot of errors and a serious structure change which we muddled through a bit before ignoring and going straight to a php8.4 version.

If looking to upgrade, I suggest ignoring the `php8.3` branch completely.
